### PR TITLE
Cache basic authentications

### PR DESCRIPTION
**Pull request by roidelapluie**
In order to improve the response time with basic authentication, cache
the results of bcrypt in memory.

This is a draft to see if that approach would be acceptable.

The code is coming from Caddy, and I have fullfiled the attribution
required by Apache-2 by adding the copyright to our headers:
https://github.com/caddyserver/caddy/commit/9a7756c6e4b4ddb945bede3ddb2dfbf241208915

Signed-off-by: Julien Pivotto <roidelapluie@inuits.eu>

### CODE REVIEW

**Comment by roidelapluie** [comment]
If the general approach is file I will add specific tests.

File: web/users.go [hunk 1]
```diff
@@ -58,7 +62,21 @@ func (u *userAuthRoundtrip) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 	user, pass, auth := r.BasicAuth()
 	if auth {
 		if hashedPassword, ok := c.Users[user]; ok {
-			if err := bcrypt.CompareHashAndPassword([]byte(hashedPassword), []byte(pass)); err == nil {
+			cacheKey := hex.EncodeToString(append(append([]byte(user), []byte(hashedPassword)...), []byte(pass)...))
```
Comment (roidelapluie) [inline web/users.go:65]: We could also sha512 this one maybe if we don't want plaintext in memory.

File: web/cache.go [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package web
+
+import (
+	weakrand "math/rand"
```
Comment (SuperQ) [inline web/cache.go:19]: Odd choice to go with this over `crypto/rand`.

Comment (roidelapluie) [reply web/cache.go:19]: It comes from caddy and I guess that we do not need more, it is just used for eviction, not for anything security related.

Comment (SuperQ) [reply web/cache.go:19]: Yea, I understand, I'm just interested in the choice.

File: web/users.go (hunk 1 above)
Comment (SuperQ) [reply web/users.go:65]: Basic auth is plain text anyway, so I'm not that concerned about it.

File: web/cache.go [hunk 3]
```diff
@@ -0,0 +1,77 @@
+// Copyright 2020 The Prometheus Authors
```
Comment (SuperQ) [inline web/cache.go:1]:
```suggestion
// Copyright 2021 The Prometheus Authors
```

Comment (roidelapluie) [reply web/cache.go]: fixed

**Comment by roidelapluie** [comment]
is this good to go?

**Comment by roidelapluie** [comment]
Hold on.

**Comment by roidelapluie** [comment]
OK now it works as expected. I have added getters and setters to release locks early.

I have also added a Lock around bcrypt so that we only run one at the same time.

**Review by SuperQ — APPROVED** [review]
LGTM

File: web/users.go [hunk 4]
```diff
@@ -40,6 +44,10 @@ type userAuthRoundtrip struct {
 	tlsConfigPath string
 	handler       http.Handler
 	logger        log.Logger
+	cache         *cache
+	// bcryptMtx is there to ensure that bcrypt.CompareHashAndPassword is run
+	// only once in parallel as this is CPU expansive.
```
Comment (beorn7) [inline web/users.go:49]: expansive → intensive?

Comment (roidelapluie) [reply web/users.go:49]: 🤦🏻 I will fix it

**Pull request merged**